src/migrations/20231126165733-goalusers.js

```diff
@@ -0,0 +1,464 @@
+            MIN(g."createdAt") "createdAt",
+            MIN(g."createdAt") "updatedAt",
+            null::JSONB "linkBack"
+          FROM "ZALGoals" zg
+          LEFT JOIN "Users" u
+          ON zg.dml_as = u.id
+          JOIN "Goals" g
+          ON zg.data_id = g.id
+          WHERE dml_as NOT IN (-1, 0) -- default and migration files
+          AND dml_type = '${dmlType}'
+          AND new_row_data -> 'name' IS NOT NULL
+          GROUP BY 1,2
+        `,
+        typeName,
+      );
+
+      const collectGoalCollaboratorsViaActivityReport = () => collectGoalCollaborators(
+        /* sql */`
+        SELECT
+          g.id "goalId",
+          (ARRAY_AGG(ar."userId" ORDER BY ar.id ASC))[1] "userId",
+          MIN(ar."createdAt") "createdAt",
+          MIN(ar."createdAt") "updatedAt",
+          null::JSONB "linkBack"
+        FROM "Goals" g
+        LEFT JOIN "ZALGoals" zg
+        ON g.id = zg.data_id
+        LEFT JOIN "ActivityReportGoals" arg
+        ON g.id = arg."goalId"
+        LEFT JOIN "ActivityReports" ar
+        ON arg."activityReportId" = ar.id
+        WHERE (zg.id IS NULL
+        OR (zg.dml_as IN (-1, 0)
+          AND zg.dml_type = 'INSERT'))
+        AND g."createdVia" IN ('activityReport')
+        GROUP BY 1
+        HAVING (ARRAY_AGG(ar."userId" ORDER BY ar.id ASC))[1] IS NOT NULL
+        AND MIN(ar."createdAt") IS NOT NULL
+        ORDER BY 1
+        `,
+        GOAL_COLLABORATORS.CREATOR,
+      );
+
+      const collectGoalCollaboratorsAsLinkers = () => collectGoalCollaborators(
+        /* sql */`
+       SELECT
+        "goalId",
+        "userId",
+        MIN("createdAt") "createdAt",
+        MAX("updatedAt") "updatedAt",
+        jsonb_build_object('activityReportIds', ARRAY_AGG(DISTINCT ar."activityReportId")) "linkBack"
+      FROM (
+        SELECT
+        arg."goalId" "goalId",
+        zarg.dml_as "userId",
+        MIN(arg."createdAt") "createdAt",
+        MAX(arg."createdAt") "updatedAt",
+          ARRAY_AGG(DISTINCT arg."activityReportId") "activityReportIds"
+        FROM "ZALActivityReportGoals" zarg
+        LEFT JOIN "ZALActivityReportGoals" zargd
+        ON zarg.data_id = zargd.data_id
+        AND zarg.id < zargd.id
+        AND zarg.dml_type = 'INSERT'
+        AND zargd.dml_type = 'DELETE'
+        JOIN "ActivityReportGoals" arg
+        ON zarg.data_id = arg.id
+        WHERE zargd.id IS NULL
+        AND zarg.dml_type = 'INSERT'
+        AND zarg.dml_as NOT IN (-1, 0)
+        GROUP BY 1,2
+        UNION
+        SELECT
+        arg."goalId",
+        ar."userId",
+        MIN(arg."createdAt") "createdAt",
+        MAX(arg."createdAt") "updatedAt",
+          ARRAY_AGG(DISTINCT arg."activityReportId") "activityReportIds"
+        FROM "ActivityReportGoals" arg
+        JOIN "ActivityReports" ar
+        ON arg."activityReportId" = ar.id
+        LEFT JOIN "ZALActivityReportGoals" zarg
+        ON arg.id = zarg.data_id
+        AND zarg.dml_type = 'INSERT'
```

Comment:
Shouldn't this one also filter out cases where the dml_as is in -1 and 0? Otherwise we'll miss cases where pre-audit log ARs get ARGs via migration.